### PR TITLE
add readout error and its mitigation

### PR DESCRIPTION
**Pull request by yutuer21**

### CODE REVIEW

File: tensorcircuit/basecircuit.py [hunk 1]
```diff
@@ -518,6 +520,8 @@ def sample(
         :param allow_state: if true, we sample from the final state
             if memory allsows, True is prefered, defaults to False
         :type allow_state: bool, optional
+        :param readouterror: readoutlist, defaults to None
```
Comment (refraction-ray) [inline tensorcircuit/basecircuit.py:523]: `readout_error`

File: tensorcircuit/basecircuit.py [hunk 2]
```diff
@@ -669,6 +696,50 @@ def sample_expectation_ps(
 
     sexpps = sample_expectation_ps
 
+    def readouterror_bs(
+        self, readouterror: Optional[Sequence[Any]], p: Optional[Any]
+    ) -> Tensor:
```
Comment (refraction-ray) [inline tensorcircuit/basecircuit.py:701]: return real

File: tensorcircuit/basecircuit.py [hunk 3]
```diff
@@ -669,6 +696,50 @@ def sample_expectation_ps(
...
+        :type readouterror: Optional[Sequence[Any]]
+        :param p: probabilities of bit string
+        :type p: Optional[Any]
+        :rtype: Tensor
+        """
+        if isinstance(readouterror, list):
+            nqubit = len(readouterror)
+            readoutlist = []
+            for i in range(nqubit):
+                readoutlist.append(
```
Comment (refraction-ray) [inline tensorcircuit/basecircuit.py:721]: jittable

File: tensorcircuit/basecircuit.py [hunk 4]
```diff
@@ -669,6 +696,50 @@ def sample_expectation_ps(
...
+            ms = [Gate(readoutlist[i]) for i in range(nqubit)]
+            p = backend.cast(p, dtypestr)
+            p = Gate(backend.reshape2(p))
+            for i in range(nqubit):
+                ms[i][1] ^ p[i]
+            nodes = ms + [p]
+            r = contractor(nodes, output_edge_order=[m[0] for m in ms]).tensor
+            p = backend.reshape(r, [-1])
+        else:
+            raise ValueError("readouterror should be a list")
```
Comment (refraction-ray) [inline tensorcircuit/basecircuit.py:740]: tensor is possible

File: tensorcircuit/basecircuit.py [hunk 5]
```diff
@@ -669,6 +696,50 @@ def sample_expectation_ps(
...
+        >>> readouterror.append([0.4,0.7])   # readout error for qubit 1, [p0|0,p1|1]
+
+
+        :param readouterror: list of readout error for each qubits.
+        :type readouterror: Optional[Sequence[Any]]
+        :param p: probabilities of bit string
+        :type p: Optional[Any]
+        :rtype: Tensor
+        """
+        if isinstance(readouterror, list):
```
Comment (refraction-ray) [inline tensorcircuit/basecircuit.py:717]: tuple, tensor should also work

File: tensorcircuit/basecircuit.py [hunk 6]
```diff
@@ -10,6 +10,7 @@
 import graphviz
 import tensornetwork as tn
 
+from . import cons
```
Comment (refraction-ray) [inline tensorcircuit/basecircuit.py:13]: delete

File: tests/test_channels.py [hunk 7]
```diff
@@ -150,3 +152,197 @@ def noisecircuitdm():
...
+    c.X(0)
+
+    value = c.sample_expectation_ps(z=[0, 1, 2])
+    valueaim = -1
+    np.testing.assert_allclose(value, valueaim, atol=1e-3)
+
+    readouterror = []
+    readouterror.append([0.9, 0.75])  # readout error of qubit 0
+    readouterror.append([0.4, 0.7])  # readout error of qubit 1
+    readouterror.append([0.7, 0.9])  # readout error of qubit 2
```
Comment (refraction-ray) [inline tests/test_channels.py:171]: tensor

File: tests/test_channels.py [hunk 8]
```diff
@@ -150,3 +152,197 @@ def noisecircuitdm():
...
+        batch=1000,
+        allow_state=True,
+        readouterror=readouterror,
+        format_="count_dict_bin",
+        random_generator=key,
+    )
+    print(bs)
+
+    # test jitble
+    def jitest(readouterror):
```
Comment (refraction-ray) [inline tests/test_channels.py:244]: jit_test

File: tests/test_channels.py [hunk 9]
```diff
@@ -150,3 +152,197 @@ def noisecircuitdm():
...
+    calsample = tc.backend.jit(jitest)
+    sampletest = calsample(readouterror)
+    print(sampletest)
+
+
+def cali_readout_circ(nqubit):
+    calicirc = []
+    for i in range(2**nqubit):
+        name = "{:0" + str(nqubit) + "b}"
+        lisbs = [int(x) for x in name.format(i)]
```
Comment (refraction-ray) [inline tests/test_channels.py:259]: bs_list

File: tensorcircuit/basecircuit.py (hunk 1 above)
Comment (yutuer21) [reply tensorcircuit/basecircuit.py:523]: revised

File: tensorcircuit/basecircuit.py (hunk 2 above)
Comment (yutuer21) [reply tensorcircuit/basecircuit.py:701]: revised

**Pull request merged**